packages/next/src/server/web/spec-extension/unstable-cache.ts

```diff
@@ -30,164 +60,245 @@ export function unstable_cache<T extends Callback>(
+      const implicitTags = addImplicitTags(store)
+
+      if (
+        // when we are nested inside of other unstable_cache's
+        // we should bypass cache similar to fetches
+        store.fetchCache !== 'force-no-store' &&
+        !store.isOnDemandRevalidate &&
+        !incrementalCache.isOnDemandRevalidate
+      ) {
+        // We attempt to get the current cache entry from the incremental cache.
+        const cacheEntry = await incrementalCache.get(cacheKey, {
+          kindHint: 'fetch',
+          revalidate: options.revalidate,
+          tags,
+          softTags: implicitTags,
+        })
+
+        if (cacheEntry && cacheEntry.value) {
+          // The entry exists and has a value
+          if (cacheEntry.value.kind !== 'FETCH') {
+            // The entry is invalid and we need a special warning
+            // @TODO why do we warn this way? Should this just be an error? How are these errors surfaced
+            // so bugs can be reported
+            // @TODO the invocation key can have sensitive data in it. we should not log this entire object
+            console.error(
+              `Invariant invalid cacheEntry returned for ${invocationKey}`
             )
+            // will fall through to generating a new cache entry below
+          } else {
+            // We have a valid cache entry so we will be returning it. We also check to see if we need
+            // to background revalidate it by checking if it is stale.
+            const cachedResponse = JSON.parse(cacheEntry.value.data.body)
+            if (cacheEntry.isStale) {
+              // In App Router we return the stale result and revalidate in the background
+              if (!store.pendingRevalidates) {
+                store.pendingRevalidates = {}
+              }
+              // We run the cache function asynchronously and save the result when it completes
+              store.pendingRevalidates[invocationKey] =
+                staticGenerationAsyncStorage
+                  .run(
+                    {
+                      ...store,
+                      // force any nested fetches to bypass cache so they revalidate
+                      // when the unstable_cache call is revalidated
+                      fetchCache: 'force-no-store',
+                      isUnstableCacheCallback: true,
+                    },
+                    cb,
+                    ...args
+                  )
+                  .then((result) => {
+                    return cacheNewResult(
+                      result,
+                      incrementalCache,
+                      cacheKey,
+                      tags,
+                      options.revalidate
+                    )
+                  })
+                  // @TODO This error handling seems wrong. We swallow the error?
+                  .catch((err) =>
+                    console.error(
+                      `revalidating cache with key: ${invocationKey}`,
```

Comment:
this could leak args to logs. Probably not safe to do. We should come up with some other way to identify a revalidation for investigation.